File: source/bindings/csharp/conversation_translator.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Threading.Tasks;
using Microsoft.CognitiveServices.Speech.Internal;
using static Microsoft.CognitiveServices.Speech.Internal.SpxExceptionThrower;
using System.Threading;
using System.Runtime.CompilerServices;
using System.Globalization;

namespace Microsoft.CognitiveServices.Speech.Transcription
{
    /// <summary>
    /// A conversation translator that enables a connected experience where participants can use their
    /// own devices to see everyone else's recognitions and IMs in their own languages. Participants
    /// can also speak and send IMs to others.
    /// Added in 1.9.0
    /// </summary>
    public sealed class ConversationTranslator : Internal.DisposableBase
    {
        internal Internal.InteropSafeHandle _nativeHandle;
        private int _activeAsyncCounter = 0;

        private InteropEvent<SessionEventArgs> _sessionStarted;
        private InteropEvent<SessionEventArgs> _sessionStopped;
        private InteropEvent<ConversationTranslationCanceledEventArgs> _canceled;
        private InteropEvent<ConversationParticipantsChangedEventArgs> _participantsChanged;
        private InteropEvent<ConversationExpirationEventArgs> _conversationExpiration;
        private InteropEvent<ConversationTranslationEventArgs> _transcribing;
        private InteropEvent<ConversationTranslationEventArgs> _transcribed;
        private InteropEvent<ConversationTranslationEventArgs> _textMessageReceived;

        private Audio.AudioConfig _audioInputKeepAlive;

        /// <summary>
        /// Creates a new instance of the Conversation Translator using the default microphone input.
        /// </summary>
        public ConversationTranslator()
            : this(Create(null, false))
        {
        }

        /// <summary>
        /// Creates a new instance of the Conversation Translator.
        /// </summary>
        /// <param name="audioConfig">Audio configuration.</param>
        /// <exception cref="ArgumentNullException">If the audioConfig is null.</exception>
        public ConversationTranslator(Audio.AudioConfig audioConfig)
            : this(Create(audioConfig, true))
        {
            _audioInputKeepAlive = audioConfig;
        }

        /// <summary>
        /// Creates a new instance.
        /// </summary>
        /// <param name="nativeHandle">The handle to the native conversation translator.</param>
        internal ConversationTranslator(Internal.InteropSafeHandle nativeHandle)
        {
            _nativeHandle = nativeHandle ?? throw new ArgumentNullException(nameof(nativeHandle));

            _sessionStarted = new InteropEvent<SessionEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_session_started_set_callback);
            _sessionStopped = new InteropEvent<SessionEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_session_stopped_set_callback);
            _canceled = new InteropEvent<ConversationTranslationCanceledEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_canceled_set_callback);
            _participantsChanged = new InteropEvent<ConversationParticipantsChangedEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_participants_changed_set_callback);
            _conversationExpiration = new InteropEvent<ConversationExpirationEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_conversation_expiration_set_callback);
            _transcribing = new InteropEvent<ConversationTranslationEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_transcribing_set_callback);
            _transcribed = new InteropEvent<ConversationTranslationEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_transcribed_set_callback);
            _textMessageReceived = new InteropEvent<ConversationTranslationEventArgs>(
                this, _nativeHandle, Internal.ConversationTranslator.conversation_translator_text_message_recevied_set_callback);

            IntPtr propertyHandle = IntPtr.Zero;
            ThrowIfFail(Internal.ConversationTranslator.conversation_translator_get_property_bag(nativeHandle, out propertyHandle));
            Properties = new PropertyCollection(propertyHandle);
        }

        /// <summary>
        /// Event that signals the start of a conversation translation session.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStarted
        {
            add { _sessionStarted?.Add(value); }
            remove { _sessionStarted.Remove(value); }
        }

        /// <summary>
        /// Event that signals the end of a conversation translation session.
        /// </summary>
        public event EventHandler<SessionEventArgs> SessionStopped
        {
            add { _sessionStopped?.Add(value); }
            remove { _sessionStopped.Remove(value); }
        }

        /// <summary>
        /// Event that signals an error with the conversation transcription, or the end of the
        /// audio stream has been reached.
        /// </summary>
        public event EventHandler<ConversationTranslationCanceledEventArgs> Canceled
        {
            add { _canceled?.Add(value); }
            remove { _canceled.Remove(value); }
        }

        /// <summary>
        /// Event that signals participants in the room have changed (e.g. a new participant joined).
        /// </summary>
        public event EventHandler<ConversationParticipantsChangedEventArgs> ParticipantsChanged
        {
            add { _participantsChanged?.Add(value); }
            remove { _participantsChanged.Remove(value); }
        }

        /// <summary>
        /// Event that signals how many more minutes are left before the conversation expires.
        /// </summary>
        public event EventHandler<ConversationExpirationEventArgs> ConversationExpiration
        {
            add { _conversationExpiration?.Add(value); }
            remove { _conversationExpiration.Remove(value); }
        }

        /// <summary>
        /// Event that signals an intermediate conversation translation result is available for a
        /// conversation participant.
        /// </summary>
        public event EventHandler<ConversationTranslationEventArgs> Transcribing
        {
            add { _transcribing?.Add(value); }
            remove { _transcribing.Remove(value); }
        }

        /// <summary>
        /// Event that signals a final conversation translation result is available for a conversation
        /// participant.
        /// </summary>
        public event EventHandler<ConversationTranslationEventArgs> Transcribed
        {
            add { _transcribed?.Add(value); }
            remove { _transcribed.Remove(value); }
        }

        /// <summary>
        /// Event that signals a translated text message from a conversation participant.
        /// </summary>
        public event EventHandler<ConversationTranslationEventArgs> TextMessageReceived
        {
            add { _textMessageReceived?.Add(value); }
            remove { _textMessageReceived.Remove(value); }
        }

        /// <summary>
        /// Gets the language name that is used for recognition.
        /// </summary>
        public string SpeechRecognitionLanguage
            => Properties.GetProperty(PropertyId.SpeechServiceConnection_RecoLanguage);

        /// <summary>
        /// Gets your participant identifier
        /// </summary>
        public string ParticipantId
            => Properties.GetProperty(PropertyId.Conversation_ParticipantId);

        /// <summary>
        /// Gets or sets the authorization token used to connect to the conversation service
        /// </summary>
        public string AuthorizationToken => Properties.GetProperty(PropertyId.SpeechServiceAuthorization_Token);

        /// <summary>
        /// Gets the collection of properties and their values defined for this <see cref="ConversationTranslator"/>.
        /// </summary>
        public PropertyCollection Properties { get; internal set; }

        /// <summary>
        /// Joins an existing conversation. You should use this method if you have created a conversation
        /// using <see cref="Conversation.CreateConversationAsync(SpeechConfig, string)"/>.
        /// </summary>
        /// <param name="conversation">The conversation to join.</param>
        /// <param name="nickname">The display name to use for the current participant.</param>
        /// <returns>An asynchronous operation.</returns>
        public Task JoinConversationAsync(Conversation conversation, string nickname)
        {
            if (conversation == null)
            {
                throw new ArgumentNullException(nameof(conversation));
            }

            return RunAsync((h) =>
            {
                using (var nickHandle = new Utf8StringHandle(nickname))
                {
                    return Internal.ConversationTranslator.conversation_translator_join(
                        h, conversation.NativeHandle, nickHandle);
                }
            });
        }

        /// <summary>
        /// Joins an existing conversation.
        /// </summary>
        /// <param name="conversationId">The unique identifier for the conversation to join.</param>
        /// <param name="nickname">The display name to use for the current participant.</param>
        /// <param name="lang">The speech language to use for the current participant.</param>
        /// <returns>An asynchronous operation.</returns>
        public Task JoinConversationAsync(string conversationId, string nickname, string lang)
        {
            if (string.IsNullOrWhiteSpace(conversationId))
            {
                throw new ArgumentException(nameof(conversationId) + " cannot be null, empty or consist only of white space");
            }
            else if (string.IsNullOrWhiteSpace(lang))
            {
                throw new ArgumentException(nameof(lang) + " cannot be null, empty or consist only of white space");
            }

            return RunAsync((h) =>
            {
                using (var idHandle = new Utf8StringHandle(conversationId))
                using (var nickHandle = new Utf8StringHandle(nickname))
                using (var langHandle = new Utf8StringHandle(lang))
                {
                    return Internal.ConversationTranslator.conversation_translator_join_with_id(
                        h, idHandle, nickHandle, langHandle);
                }
            });
        }

        /// <summary>
        /// Starts sending audio to the conversation service for speech recognition and translation. You
        /// should subscribe to the <see cref="Transcribing"/>, and <see cref="Transcribed"/> events to
        /// receive conversation translation results for yourself, and other participants in the
        /// conversation.
        /// </summary>
        /// <returns>An asynchronous operation.</returns>
        public Task StartTranscribingAsync()
            => RunAsync(Internal.ConversationTranslator.conversation_translator_start_transcribing);

        /// <summary>
        /// Stops sending audio to the conversation service. You will still receive <see cref="Transcribing"/>,
        /// and <see cref="Transcribed"/> events for other participants in the conversation.
        /// </summary>
        /// <returns>An asynchronous operation.</returns>
        public Task StopTranscribingAsync()
            => RunAsync(Internal.ConversationTranslator.conversation_translator_stop_transcribing);

        /// <summary>
        /// Sends an instant message to all participants in the conversation. This instant message
        /// will be translated into each participant's text language.
        /// </summary>
        /// <param name="message">The message to send.</param>
        /// <returns>An asynchronous operation.</returns>
        public Task SendTextMessageAsync(string message)
        {
            if (string.IsNullOrWhiteSpace(message))
            {
                throw new ArgumentException(nameof(message) + " cannot be null, empty or consist only of white space");
            }

            return RunAsync((h) =>
            {
                using (var messageHandle = new Utf8StringHandle(message))
                {
                    return Internal.ConversationTranslator.conversation_translator_send_text_message(
                        h, messageHandle);
                }
            });
        }

        /// <summary>
        /// Sets the Cognitive Speech authorization token that will be used for connecting to the server.
        /// </summary>
        /// <param name="authToken">The authorization token.</param>
        /// <param name="region">(Optional) The Azure region for this token.</param>
        public void SetAuthorizationToken(string authToken, string region = null)
        {
            if (string.IsNullOrWhiteSpace(authToken))
            {
                throw new ArgumentException(nameof(authToken) + " cannot be null, empty or consist only of white space");
            }

            using (var authTokenHandle = new Utf8StringHandle(authToken))
            using (var regionHandle = new Utf8StringHandle(region))
            {
                ThrowIfFail(Internal.ConversationTranslator.conversation_translator_set_authorization_token(_nativeHandle, authTokenHandle, regionHandle));
            }
        }

        /// <summary>
        /// Leave the current conversation. After this is called, you will no longer receive any events.
        /// </summary>
        /// <returns>An asynchronous operation.</returns>
        public Task LeaveConversationAsync()
            => RunAsync(Internal.ConversationTranslator.conversation_translator_leave);

        /// <summary>
        /// Disposes of the object
        /// </summary>
        /// <param name="disposeManaged">True to dispose managed resources</param>
        /// <returns></returns>
        protected override void Dispose(bool disposeManaged)
        {
            if (disposeManaged)
            {
                if (_activeAsyncCounter > 0)
                {
                    throw new InvalidOperationException("Cannot dispose a conversation translator while an async action is running. Await async actions to avoid unexpected disposals.");
                }

                _sessionStarted?.Dispose();
                _sessionStopped?.Dispose();
                _canceled?.Dispose();
                _participantsChanged?.Dispose();
                _conversationExpiration?.Dispose();
                _transcribing?.Dispose();
                _transcribed?.Dispose();
                _textMessageReceived?.Dispose();

                // This will make Properties unaccessible.
                Properties.Close();
            }
            _audioInputKeepAlive = null;
            _nativeHandle?.Dispose();
        }

        private static InteropSafeHandle Create(Audio.AudioConfig audioConfig, bool validateConfig)
        {
            if (validateConfig && audioConfig == null)
            {
                throw new ArgumentNullException(nameof(audioConfig));
            }

            IntPtr handle = Internal.ConversationTranslator.SPXHANDLE_INVALID;
            InteropSafeHandle audioConfigHandle = audioConfig == null
                ? new InteropSafeHandle(new IntPtr(-1), null)
                : audioConfig.configHandle;

            ThrowIfFail(Internal.ConversationTranslator.conversation_translator_create_from_config(
                out handle, audioConfigHandle));

            if (audioConfig != null)
            {
                GC.KeepAlive(audioConfig);
            }

            return new InteropSafeHandle(handle, Internal.ConversationTranslator.conversation_translator_handle_release);
        }

        private Task RunAsync(Func<InteropSafeHandle, IntPtr> method, [CallerMemberName]string caller = null, [CallerLineNumber]int line = 0)
        {
            return Task.Run(() =>
            {
                CheckDisposed();
                ThrowIfNull(_nativeHandle);

                try
                {
                    Interlocked.Increment(ref _activeAsyncCounter);
                    ThrowIfFail(method(_nativeHandle));
                }
                catch (Exception ex)
                {
                    string source = string.Format(CultureInfo.InvariantCulture, "[{0}:{1}] {2}", caller, line, method?.Method);

                    // add information to the exception about who called this method, as well the function that failed
                    ex.Source = source;

                    // log to debug output
                    string debugLog = string.Format(CultureInfo.InvariantCulture, "{0} - {1}: {2}", ex.Source, ex.GetType().Name, ex.Message);
                    System.Diagnostics.Debug.WriteLine(debugLog);

                    // finally rethrow the original exception
                    throw;
                }
                finally
                {
                    Interlocked.Decrement(ref _activeAsyncCounter);
                }
            });
        }
    }
}

File: source/bindings/java/com/microsoft/cognitiveservices/speech/transcription/ConversationTranslator.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license
// information.
//
package com.microsoft.cognitiveservices.speech.transcription;

import com.microsoft.cognitiveservices.speech.PropertyCollection;
import com.microsoft.cognitiveservices.speech.PropertyId;
import com.microsoft.cognitiveservices.speech.SessionEventArgs;
import com.microsoft.cognitiveservices.speech.audio.AudioConfig;
import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.EventHandlerImpl;
import com.microsoft.cognitiveservices.speech.util.IntRef;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;
import com.microsoft.cognitiveservices.speech.util.SafeHandleType;
import java.io.Closeable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A conversation translator that enables a connected experience where participants can use their
 * own devices to see everyone else's recognitions and IMs in their own languages. Participants can
 * also speak and send IMs to others. Added in version 1.12.0
 */
public class ConversationTranslator implements Closeable {

    /*! \cond PROTECTED */

    protected static ExecutorService s_executorService;

    protected AtomicInteger eventCounter = new AtomicInteger(0);
    protected Integer backgroundAttempts = 0;
    /*! \endcond */

    static {
        s_executorService = Executors.newCachedThreadPool();
    }

    /** Creates a new instance of the Conversation Translator using the default microphone input. */
    public ConversationTranslator() {
        this(null);
    }

    /**
     * Creates a new instance of the Conversation Translator.
     *
     * @param audioConfig Audio configuration.
     */
    public ConversationTranslator(AudioConfig audioConfig) {
        translatorHandle = new SafeHandle(0, SafeHandleType.ConversationTranslator);
        // Note: Since ais is optional, no test for null reference
        audioInputKeepAlive = audioConfig;
        if (audioConfig == null) {
            Contracts.throwIfFail(createConversationTranslatorFromConfig(translatorHandle, null));
        } else {
            Contracts.throwIfFail(
                    createConversationTranslatorFromConfig(translatorHandle, audioConfig.getImpl()));
        }
        initialize();
    }

    /**
     * The collection or properties and their values defined for this ConversationTranslator.
     *
     * @return The collection or properties and their values defined for this
     *     ConversationTranslator.
     */
    public PropertyCollection getProperties() {
        return propertyHandle;
    }

    /**
     * Gets the language name that is used for recognition.
     *
     * @return The language name that is used for recognition.
     */
    public String getSpeechRecognitionLanguage() {
        return propertyHandle.getProperty(PropertyId.SpeechServiceConnection_RecoLanguage);
    }

    /**
     * Joins an existing conversation. You should use this method if you have created a conversation
     * using CreateConversationAsync(SpeechConfig, String).
     *
     * @param conversation The Conversation to join.
     * @param nickname The display name to use for the current participant.
     * @return A task representing the join operation
     */
    public Future<Void> joinConversationAsync(Conversation conversation, final String nickname) {
        final ConversationTranslator thisReco = this;
        final Conversation thisConv = conversation;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.joinConversation(
                                                translatorHandle, thisConv.getImpl(), nickname));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Joins an existing conversation. You should use this method if you have created a conversation
     * using CreateConversationAsync(SpeechConfig, String).
     *
     * @param conversationId The unique identifier for the conversation to join.
     * @param nickname The display name to use for the current participant.
     * @param language The speech language to use for the current participant.
     * @return A task representing the join operation
     */
    public Future<Void> joinConversationAsync(
            final String conversationId, final String nickname, final String language) {
        final ConversationTranslator thisReco = this;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.joinConversationId(
                                                translatorHandle, conversationId, nickname, language));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Leave the current conversation. After this is called, you will no longer receive any events.
     *
     * @return A task representing the leave operation
     */
    public Future<Void> leaveConversationAsync() {
        final ConversationTranslator thisReco = this;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.leaveConversation(translatorHandle));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Sends an instant message to all participants in the conversation. This instant message will
     * be translated into each participant's text language.
     *
     * @param message The message to send.
     * @return A task representing the send message operation
     */
    public Future<Void> sendTextMessageAsync(final String message) {
        final ConversationTranslator thisReco = this;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.sendText(translatorHandle, message));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Starts sending audio to the conversation service for speech recognition and translation. You
     * should subscribe to the Transcribing, and Transcribed events to receive conversation
     * translation results for yourself, and other participants in the conversation.
     * @return A task representing the transcribe operation
     */
    public Future<Void> startTranscribingAsync() {
        final ConversationTranslator thisReco = this;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.startTranscribing(translatorHandle));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Stops sending audio to the conversation service. You will still receive Transcribing, and
     * Transcribed events for other participants in the conversation.
     * @return A task representing the transcribe operation
     */
    public Future<Void> stopTranscribingAsync() {
        final ConversationTranslator thisReco = this;

        return s_executorService.submit(
                new java.util.concurrent.Callable<Void>() {
                    public Void call() {
                        Runnable runnable =
                                new Runnable() {
                                    public void run() {
                                        Contracts.throwIfFail(thisReco.stopTranscribing(translatorHandle));
                                    }
                                };
                        thisReco.doAsyncRecognitionAction(runnable);
                        return null;
                    }
                });
    }

    /**
     * Event that signals an error with the conversation translation, or the end of the audio
     * stream has been reached.
     */
    public final EventHandlerImpl<ConversationTranslationCanceledEventArgs> canceled =
            new EventHandlerImpl<ConversationTranslationCanceledEventArgs>(eventCounter);

    /** Event that signals how many more minutes are left before the conversation expires. */
    public final EventHandlerImpl<ConversationExpirationEventArgs> conversationExpiration =
            new EventHandlerImpl<ConversationExpirationEventArgs>(eventCounter);

    /** Event that signals participants in the room have changed (e.g. a new participant joined). */
    public final EventHandlerImpl<ConversationParticipantsChangedEventArgs> participantsChanged =
            new EventHandlerImpl<ConversationParticipantsChangedEventArgs>(eventCounter);

    /** Event that signals the start of a conversation translation session. */
    public final EventHandlerImpl<SessionEventArgs> sessionStarted =
            new EventHandlerImpl<SessionEventArgs>(eventCounter);

    /** Event that signals the end of a conversation translation session. */
    public final EventHandlerImpl<SessionEventArgs> sessionStopped =
            new EventHandlerImpl<SessionEventArgs>(eventCounter);

    /** Event that signals a translated text message from a conversation participant. */
    public final EventHandlerImpl<ConversationTranslationEventArgs> textMessageReceived =
            new EventHandlerImpl<ConversationTranslationEventArgs>(eventCounter);

    /**
     * Event that signals a final conversation translation result is available for a conversation
     * participant.
     */
    public final EventHandlerImpl<ConversationTranslationEventArgs> transcribed =
            new EventHandlerImpl<ConversationTranslationEventArgs>(eventCounter);

    /**
     * Event that signals an intermediate conversation translation result is available for a
     * conversation participant.
     */
    public final EventHandlerImpl<ConversationTranslationEventArgs> transcribing =
            new EventHandlerImpl<ConversationTranslationEventArgs>(eventCounter);

    /**
     * Dispose of associated resources. Note: close() must be called in order to relinquish
     * underlying resources held by the object.
     */
    @Override
    public void close() {
        synchronized (translatorLock) {
            if (activeAsyncRecognitionCounter != 0) {
                throw new IllegalStateException(
                        "Cannot dispose a conversationTranslator while async transcription is running. Await async recognitions to avoid unexpected disposals.");
            }
            dispose(true);
        }
    }

    /*! \cond INTERNAL */
    /**
     * Returns the internal conversation translator instance
     * @return The internal conversation translator instance
     */
    public SafeHandle getImpl() {
        return translatorHandle;
    }
    /*! \endcond */

    private void dispose(final boolean disposing) {
        if (disposed) {
            return;
        }

        if (disposing) {
            if (this.eventCounter.get() != 0 && backgroundAttempts <= 50) {
                // There is an event callback in progress, closing while in an event call results in
                // SWIG problems, so
                // spin a thread to try again in 500ms and return.
                getProperties()
                        .getProperty(
                                "Backgrounding release "
                                        + backgroundAttempts.toString()
                                        + " "
                                        + this.eventCounter.get(),
                                "");
                Thread t =
                        new Thread(
                                new Runnable() {
                                    @Override
                                    public void run() {
                                        try {
                                            Thread.sleep(500 * ++backgroundAttempts);
                                            dispose(disposing);
                                        } catch (Exception e) {
                                        }
                                    }
                                });
                t.start();
            } else {

                if (propertyHandle != null) {
                    propertyHandle.close();
                    propertyHandle = null;
                }
                if (translatorHandle != null) {
                    translatorHandle.close();
                    translatorHandle = null;
                }
                conversationTranslatorObjects.remove(this);
                audioInputKeepAlive = null;
                disposed = true;
            }
        }
    }

    private void doAsyncRecognitionAction(Runnable recoImplAction) {
        synchronized (translatorLock) {
            activeAsyncRecognitionCounter++;
        }
        if (disposed) {
            throw new IllegalStateException(this.getClass().getName());
        }
        try {
            recoImplAction.run();
        } finally {
            synchronized (translatorLock) {
                activeAsyncRecognitionCounter--;
            }
        }
    }

    private void initialize() {
        final ConversationTranslator _this = this;

        this.sessionStarted.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(
                                sessionStartedSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.sessionStopped.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(
                                sessionStoppedSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.canceled.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(canceledSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.conversationExpiration.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(
                                conversationExpireSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.participantsChanged.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(
                                participantsChangedSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.transcribing.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(transcribingSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.transcribed.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(transcribedSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        this.textMessageReceived.updateNotificationOnConnected(
                new Runnable() {
                    @Override
                    public void run() {
                        conversationTranslatorObjects.add(_this);
                        Contracts.throwIfFail(textMessageSetCallback(_this.translatorHandle.getValue()));
                    }
                });

        IntRef propHandle = new IntRef(0);
        Contracts.throwIfFail(getPropertyBagFromHandle(_this.translatorHandle, propHandle));
        propertyHandle = new PropertyCollection(propHandle);
    }

    private void sessionStartedEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            SessionEventArgs arg = new SessionEventArgs(eventArgs, true);
            EventHandlerImpl<SessionEventArgs> handler = this.sessionStarted;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }
        } catch (Exception ex) {
        }
    }

    private void sessionStoppedEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            SessionEventArgs arg = new SessionEventArgs(eventArgs, true);
            EventHandlerImpl<SessionEventArgs> handler = this.sessionStopped;
            if (handler != null) {
                handler.fireEvent(this, arg);
            }
        } catch (Exception ex) {
        }
    }

    private void canceledEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationTranslationCanceledEventArgs resultEventArg =
                    new ConversationTranslationCanceledEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationTranslationCanceledEventArgs> handler = this.canceled;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    private void participantsChangedEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationParticipantsChangedEventArgs resultEventArg =
                    new ConversationParticipantsChangedEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationParticipantsChangedEventArgs> handler =
                    this.participantsChanged;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    private void conversationExpireEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationExpirationEventArgs resultEventArg =
                    new ConversationExpirationEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationExpirationEventArgs> handler = this.conversationExpiration;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    private void transcribingEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationTranslationEventArgs resultEventArg =
                    new ConversationTranslationEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationTranslationEventArgs> handler = this.transcribing;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    private void transcribedEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationTranslationEventArgs resultEventArg =
                    new ConversationTranslationEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationTranslationEventArgs> handler = this.transcribed;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    private void textMessageEventCallback(long eventArgs) {
        try {
            Contracts.throwIfNull(this, "conversationTranslator");
            if (this.disposed) {
                return;
            }
            ConversationTranslationEventArgs resultEventArg =
                    new ConversationTranslationEventArgs(eventArgs, true);
            EventHandlerImpl<ConversationTranslationEventArgs> handler = this.textMessageReceived;
            if (handler != null) {
                handler.fireEvent(this, resultEventArg);
            }
        } catch (Exception ex) {
        }
    }

    /**
     * This is used to keep any instance of this class alive that is subscribed to downstream
     * events.
     */
    static java.util.Set<ConversationTranslator> conversationTranslatorObjects =
            java.util.Collections.synchronizedSet(new java.util.HashSet<ConversationTranslator>());

    private SafeHandle translatorHandle = null;

    private final native long createConversationTranslatorFromConfig(
            SafeHandle translatorHandle, SafeHandle audioConfigHandle);

    private final native long getPropertyBagFromHandle(SafeHandle translatorHandle, IntRef propertyHandle);

    private final native long sessionStartedSetCallback(long translatorHandle);

    private final native long sessionStoppedSetCallback(long translatorHandle);

    private final native long participantsChangedSetCallback(long translatorHandle);

    private final native long transcribingSetCallback(long translatorHandle);

    private final native long transcribedSetCallback(long translatorHandle);

    private final native long textMessageSetCallback(long translatorHandle);

    private final native long conversationExpireSetCallback(long translatorHandle);

    private final native long canceledSetCallback(long translatorHandle);

    private final native long joinConversation(
            SafeHandle translatorHandle, SafeHandle conversationHandle, String nickname);

    private final native long joinConversationId(
            SafeHandle translatorHandle, String conversationId, String nickname, String language);

    private final native long leaveConversation(SafeHandle translatorHandle);

    private final native long sendText(SafeHandle translatorHandle, String text);

    private final native long startTranscribing(SafeHandle translatorHandle);

    private final native long stopTranscribing(SafeHandle translatorHandle);

    private final Object translatorLock = new Object();
    private PropertyCollection propertyHandle = null;
    private boolean disposed = false;
    private int activeAsyncRecognitionCounter = 0;
    private AudioConfig audioInputKeepAlive = null;
}
